strawberryfields/backends/fockbackend/circuit.py

```diff
@@ -135,6 +113,249 @@ def reset(self, pure=None, cutoff_dim=None, num_subsystems=None):
         else:
             self._state = ops.vacuumStateMixed(self._num_modes, self._trunc)
 
+    def apply_gate_BLAS(self, mat, modes, **kwargs):
+        """Applies a specified gate to the state.
+
+        Gate application based on custom indexing and matrix multiplication.
+        Assumes the input matrix has shape (out1, in1, ...).
+
+        This implementation uses indexing and BLAS. As per stack overflow,
+        einsum doesn't actually use BLAS but rather a c implementation. In theory
+        if reshaping is efficient this should be faster.
+
+        Args:
+            mat (array[complex]): The numeric operator to be applied to the state, of shape `[trunc]*(2*n)`
+            modes (list[int]): The list of modes to which the operator is applied on
+
+        Keyword args:
+            state (array[complex]): The state that the operator is applied to
+            pure (bool): Whether the state is pure or mixed
+            n (int): The total number of modes
+
+        Returns:
+            array[complex]: The state after application of the two-mode operation
+        """
+        # pylint: disable=too-many-branches
+
+        state = kwargs.get("state", self._state)
+        pure = kwargs.get("pure", self._pure)
+        n = kwargs.get("n", self._num_modes)
+
+        size = len(modes)
+        dim = self._trunc**size
+        stshape = [self._trunc for i in range(size)]
+
+        # Apply the following matrix transposition:
+        # |m1><m1| |m2><m2| ... |mn><mn| -> |m1>|m2>...|mn><m1|<m2|...<mn|
+        transpose_list = [2*i for i in range(size)] + [2*i + 1 for i in range(size)]
+        matview = np.transpose(mat, transpose_list).reshape((dim, dim))
+
+        # checks if matview is diagonal and, if so, use faster contractions
+        diag = np.all(matview == np.diag(np.diagonal(matview)))
+
+        if pure:
+            if n == 1:
+                if diag:
+                    mat_diag = mat.diagonal()
+                    return np.multiply(mat_diag, state)
+                return np.dot(mat, state)
+
+            # Transpose the state into the following form:
+            # |psi> |mode[0]> |mode[1]> ... |mode[n]>
+            transpose_list = [i for i in range(n) if not i in modes] + modes
+            view = np.transpose(state, transpose_list)
+
+            # Apply matrix to each substate
+            ret = np.zeros([self._trunc for i in range(n)], dtype=def_type)
+            for i in product(*([range(self._trunc) for j in range(n - size)])):
+                if diag:
+                    mat_diag = matview.diagonal()
+                    ret[i] = np.multiply(mat_diag, view[i].ravel()).reshape(stshape)
+                else:
+                    ret[i] = np.dot(matview, view[i].ravel()).reshape(stshape)
+
+            # "untranspose" the return matrix ret
+            untranspose_list = [0] * len(transpose_list)
+            for i in range(len(transpose_list)): # pylint: disable=consider-using-enumerate
```

Comment:
Just curious, was this pylint statement in the original version? I'm wondering why we don't instead just change it to
```python
for i, t in enumerate(transpose_list):
    untranspose_list[t] = i
```

Reply:
That was there before (I didn't add it). I'm not sure why enumerate isn't used instead; that would be much cleaner. Should I try to change it as part of this PR?